### PR TITLE
(Closes #298) add support for mold and tests

### LINKED ISSUE #298: Allocate with mold= parameter
This is valid Fortran, but not accepted by fparser:
```
subroutine a(b1)
   integer, dimension(:,:), allocatable :: b1
   integer, dimension(:,:), allocatable :: a1
   allocate(a1, mold=b1)
end subroutine a
```
Supporting the `mold` argument makes it easier to create the kernel-extraction driver (since we can then allocate variables without having to analyse the number of dimensions). It's not urgent, since I have a work around.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

File: doc/fparser2.rst [hunk 1]
```diff
@@ -42,8 +42,8 @@ Fortran 2003. This is implemented in the Fortran2003.py `file`__ and
 contains an entirely separate parser to fparser1 that includes rules
 for Fortran 2003 syntax. Support for Fortran 2008 is being added in
 the Fortran2008.py `file`__ which extends the Fortran2003 rules
-appropriately. At this time fparser2 supports submodules, co-arrays
-and the continguous keyword in Fortran2008.
+appropriately. At this time fparser2 supports submodules, co-arrays,
+the 'mold' argument to allocate and the continguous keyword in Fortran2008.
```
Comment (sergisiso) [inline doc/fparser2.rst:46]: s/continguous/contiguous/

File: src/fparser/two/Fortran2008.py [hunk 2]
```diff
@@ -83,13 +83,18 @@
     SeparatorBase,
     Type_Declaration_StmtBase,
     StmtBase,
+    NoMatchError,
```
Comment (sergisiso) [inline src/fparser/two/Fortran2008.py:86]: It's not used in this file. Is there a reason to import it here?

File: src/fparser/two/Fortran2008.py [hunk 3]
```diff
@@ -719,6 +726,42 @@ class Do_Term_Action_Stmt(Do_Term_Action_Stmt_2003):  # R826
...
+    Extends the Fortran2003 version of this class by updating the keyword
+    pairs (used in match) with support for MOLD.
+
+    """
+
+    _keyword_pairs = [
+        ("STAT", Stat_Variable),
+        ("ERRMSG", Errmsg_Variable),
+        ("SOURCE", Source_Expr),
+        ("MOLD", Source_Expr),
```
Comment (sergisiso) [inline src/fparser/two/Fortran2008.py:746]: Even if it is the same in practice, should this match Expr directly? Source_Expr is unrelated to MOLD

**Review by sergisiso — CHANGES_REQUESTED** [review]
It looks good @arporter , just a few small comments and bring it to master.

File: doc/fparser2.rst (hunk 1 above)
Comment (arporter) [reply doc/fparser2.rst:46]: Oops.

File: src/fparser/two/Fortran2008.py (hunk 2 above)
Comment (arporter) [reply src/fparser/two/Fortran2008.py:86]: I removed it and nothing broke so apparently not :-)

File: src/fparser/two/Fortran2008.py (hunk 3 above)
Comment (arporter) [reply src/fparser/two/Fortran2008.py:746]: The standard specifies `Source_Expr` so I went with that.

**Comment by arporter** [comment]
Ready for another look now @sergisiso

**Review by sergisiso — APPROVED** [review]
@arporter All suggestions addressed and all tests still pass. This is ready to merge

**Pull request merged**